hpx/parallel/executors/executor_parameters.hpp

```diff
@@ -0,0 +1,262 @@
+//  Copyright (c) 2016 Marcin Copik
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+/// \file parallel/executors/executor_parameters.hpp
+
+#if !defined(HPX_PARALLEL_EXECUTOR_PARAMETERS)
+#define HPX_PARALLEL_EXECUTOR_PARAMETERS
+
+#include <hpx/config.hpp>
+#include <hpx/traits/is_executor_parameters.hpp>
+#include <hpx/parallel/config/inline_namespace.hpp>
+#include <hpx/parallel/executors/executor_traits.hpp>
+#include <hpx/util/always_void.hpp>
+
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include <cstdarg>
+#include <functional>
+
+#include <boost/ref.hpp>
+namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v3)
+{
+    namespace detail {
+
+        template<template<typename> class Condition, typename Arg,
+            typename Enable = void>
+        struct counter_increment
+        {
+            constexpr static int value = 0;
+        };
+
+        template<template<typename> class Condition, typename Arg>
+        struct counter_increment<Condition, Arg,
+            typename std::enable_if< Condition<Arg>::value>::type >
+        {
+            constexpr static int value = 1;
+        };
+
+        template<template<typename> class Condition, typename ...Args>
+        struct parameter_type_counter;
+
+        template<template<typename> class Condition>
+        struct parameter_type_counter<Condition>
+        {
+            constexpr static int value = 0;
+        };
+
+        /// Return the number of parameters which have been derived from type T.
+        /// Useful for checking possible duplicates or ensuring that all passed types
+        /// are indeed executor parameters.
+        ///
+        /// \param T                [in] Parameter to look for
+        /// \param Parameters       [in] The executor object to use.
+        ///
+        //
+        template<template<typename> class Condition, typename Arg1, typename ...Args>
+        struct parameter_type_counter<Condition, Arg1, Args...>
+        {
+            constexpr static int value =
+                parameter_type_counter<Condition, Args...>::value +
+                counter_increment<Condition, Arg1>::value;
+        };
+
+        template<typename T>
+        struct unwrapper : T
+        {
+            unwrapper() {}
+
+            template<typename U>
+            unwrapper(U && u) : T(u) {}
+        };
+
+        template<typename T>
+        struct unwrapper< boost::reference_wrapper<T> >
+        {
+            template<typename WrapperType>
+            unwrapper(WrapperType && wrap) : wrap(wrap) {}
+
+            template<typename Executor, typename U = T, typename std::enable_if<
+                is_executor_parameters_chunk_size<U>::value >::type* = nullptr
+                >
```

Comment:
The test has been already implemented in HPX.

It seems I'm no setting a precedence here:

> grep -e "nullptr" -nrw .
> ./hpx/runtime/threads/policies/throttle_queue_scheduler.hpp:117:            nanosleep(&tim, nullptr);
> ./hpx/compute/vector.hpp:140:            if(static_cast<T*>(data_) != nullptr)
> ./hpx/compute/vector.hpp:156:            if(static_cast<T*>(data_) != nullptr)
> ./hpx/compute/detail/iterator.hpp:56:          : p_(nullptr)
> ./hpx/compute/detail/iterator.hpp:57:          , target_(nullptr)
> ./hpx/compute/cuda/target_ptr.hpp:43:          : p_(nullptr)
> ./hpx/compute/cuda/target_ptr.hpp:44:          , tgt_(nullptr)

Grep done a one recent checkout from hpx's master. I'm willing to add this one liner, should I change those instances too? Should I make another pull request - it's a small change, but completely unrelated to variadic executor parameters.